### PR TITLE
Fix flaky test test_gluon.test_hybrid_static_memory_switching

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
This is to fix the flaky test https://github.com/apache/incubator-mxnet/issues/11171
There was a race condition when we invalidate MKLDNN memory in NDArray. The original code invalidates MKLDNN memory in the memory planning phase (AsArray) as well, which is called outside the threaded engine. The static memory allocation mode of CachedOp shares the same NDArrays in different model executions. Therefore, the NDArrays may still be used in the threaded engine from the previous model execution, while the same CachedOp is executed for the next execution and the same NDArrays are used in memory planning. With MKLDNN enabled, NDArrays were modified in the memory planning phase, which leads to race condition.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by ZhennanQin** [comment]
Hi Zhengda,

After applying this PR with recent master, 3 unit tests failed when enabling MKLDNN. They all failed with

mxnet.base.MXNetError: [20:22:53] src/ndarray/ndarray.cc:706: Check failed: !IsMKLDNNData() We can't generate TBlob for MKLDNN data. Please use Reorder2Default() to generate a new NDArray first

I also have questions for this change. Can you explain why can't we do invalidate for those cases? Why they work well in dynamic allocation while don't in static allocation? I suspect there're some bugs between static allocation and threaded engine, and this bug doesn't relate to MKLDNN only. Because when I ran this test many times, I saw a lot double free and memory corruption without mkldnn stack.

**Comment by zheng-da** [comment]
@ZhennanQin I described how race condition happens. It occurs in a special case and may not be related to your problem.

**Comment by ZhennanQin** [comment]
Thanks for your declaration, I get the problem now. If my understanding is correct, we should avoid  invalidating MKLDNN memory in memory planning phase. Can we add some check or at least some comments for InvalidateMKLDNNData() to prevent same problem happens?(developers should carefully use InvalidateMKLDNNData() after enabling static memory allocation, they should be noticed that) Or, can static memory allocation do more sanity check to capture this kind of issue early instead of random crash?

**Comment by zheng-da** [comment]
this is a general problem. That is, we shouldn't modify NDArray outside the threaded engine. I don't think we should treat InvalidateMKLDNNData() so specially.
InvalidateMKLDNNData() will be removed once we move MKLDNN to subgraphs.

**Comment by ZhennanQin** [comment]
No questions from my side. LGTM.

File: tests/python/unittest/test_gluon.py [hunk 1]
```diff
@@ -1187,17 +1187,19 @@ def check_hybrid_static_memory_switching(**kwargs):
 
     x = mx.nd.random.uniform(shape=(4, 3, 32, 32))
     net(x)
+    x.attach_grad()
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tests/python/unittest/test_gluon.py]: without it, it seems the backward computation couldn't proceed.

[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply tests/python/unittest/test_gluon.py]: Couldn't proceed how? Usually we don't need to compute the grad w.r.t input x.

[the remark this answers is not in the document]
Comment (zheng-da) [reply tests/python/unittest/test_gluon.py]: what i mean is that without this modification, i don't see operators executed in the threaded engine. i don't know why. you can give it a try.

**Pull request merged**